File: UI/propertease_admin/lib/models/property_reservation.dart
import 'package:json_annotation/json_annotation.dart';
import 'package:propertease_admin/models/application_user.dart';

import 'property.dart';

part 'property_reservation.g.dart';

@JsonSerializable()
class PropertyReservation {
  int? id;
  DateTime? createdAt;
  DateTime? modifiedAt;
  int? totalRecordsCount;
  bool? isDeleted;
  Property? property;
  int? propertyId;
  int? clientId;
  int? renterId;
  int? numberOfGuests;
  DateTime? dateOfOccupancyStart;
  DateTime? dateOfOccupancyEnd;
  int? numberOfDays;
  int? numberOfMonths;
  double? totalPrice;
  bool? isMonthly;
  bool? isDaily;
  bool? isActive;
  int? status;
  String? reservationNumber;
  ApplicationUser? client;
  String? description;
  String? cancellationReason;
  DateTime? cancelledAt;
  String? cancelledByName;
  DateTime? confirmedAt;
  String? confirmedByName;

  PropertyReservation({
    this.id = 0,
    this.totalRecordsCount = 0,
    this.isDeleted = false,
    this.property,
    this.propertyId = 0,
    this.clientId = 1,
    this.renterId,
    this.numberOfGuests = 0,
    this.dateOfOccupancyStart,
    this.dateOfOccupancyEnd,
    this.numberOfDays = 0,
    this.numberOfMonths = 0,
    this.totalPrice = 0,
    this.isMonthly = true,
    this.isDaily = false,
    this.isActive = true,
    this.status,
    this.reservationNumber = '',
    this.client,
    this.description,
    this.cancellationReason,
    this.cancelledAt,
    this.cancelledByName,
    this.confirmedAt,
    this.confirmedByName,
  });

  factory PropertyReservation.fromJson(Map<String, dynamic> json) =>
      _$PropertyReservationFromJson(json);

  Map<String, dynamic> toJson() => _$PropertyReservationToJson(this);
}

File: PropertEase.Core/Entities/PropertyReservation.cs
using PropertEase.Core.Entities.Base;
using PropertEase.Core.Entities.Identity;
using PropertEase.Core.Enumerations;

namespace PropertEase.Core.Entities
{
    public class PropertyReservation : BaseEntity
    {
        public Property Property { get; set; }
        public string ReservationNumber { get; set; }

        public int PropertyId { get; set; }
        public ApplicationUser Renter { get; set; }
        public int RenterId { get; set; }
        public ApplicationUser Client { get; set; }
        public int ClientId { get; set; }
        public int NumberOfGuests { get; set; }
        public DateTime DateOfOccupancyStart { get; set; }
        public DateTime DateOfOccupancyEnd { get; set; }
        public int NumberOfDays { get; set; }
        public int NumberOfMonths { get; set; }
        public double TotalPrice { get; set; }
        public bool IsMonthly { get; set; }
        public bool IsDaily { get; set; }
        public string? Description { get; set; }

        // Pending Confirmed Completed / Cancelled.
        public ReservationStatus Status { get; set; } = ReservationStatus.Pending;

        // Audit trail 

        public int? ConfirmedById { get; set; }
        public DateTime? ConfirmedAt { get; set; }
        public int? CancelledById { get; set; }
        public DateTime? CancelledAt { get; set; }
        public string? CancellationReason { get; set; }


        [System.ComponentModel.DataAnnotations.Schema.NotMapped]
        public bool IsActive
        {
            get => Status == ReservationStatus.Confirmed || Status == ReservationStatus.Completed;
            set => Status = value ? ReservationStatus.Confirmed : ReservationStatus.Cancelled;
        }
    }
}
